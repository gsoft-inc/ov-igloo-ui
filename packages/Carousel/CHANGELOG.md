# @igloo-ui/carousel

## 0.1.2

### Patch Changes

- a41e1d4: Fix: moved destructuring of props to function parameters so that storybook will automatically show the default value in the props table. It was done in Modal because it needed to send props to useDialog. In this case, @default annotation was used in comments.

## 0.1.1

### Patch Changes

- 386e7cd: Reduced the padding for the modal and dialog components. Also added steps and buttons to the modal. The carousel just had a minor style update on the secondary button.

## 0.1.0

### Minor Changes

- 91c219b: Inital release of the carousel component
