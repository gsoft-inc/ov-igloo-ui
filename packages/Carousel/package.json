{
  "name": "@igloo-ui/carousel",
  "version": "0.0.0",
  "main": "dist/Carousel.js",
  "module": "dist/Carousel.js",
  "types": "dist/Carousel.d.ts",
  "styles": "dist/carousel.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Vicky Comeau vicky.comeau@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/tokens": "^2.0.0",
    "classnames": "^2.3.2"
  },
  "devDependencies": {
    "@igloo-ui/button": "^0.3.7"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
