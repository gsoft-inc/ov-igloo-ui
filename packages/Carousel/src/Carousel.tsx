import * as React from 'react';
import cx from 'classnames';

import './carousel.scss';

export interface CarouselProps extends React.ComponentProps<'div'> {
  /** Contains the slides */
  children: React.ReactNode;
  /** Add a specific class to the carousel component */
  className?: string;
  /** The current slide number starting at 0 */
  currentSlide?: number;
  /** Add a data-test tag for automated tests */
  dataTest?: string;
  /** Event called when the page is changed */
  onPageChange?: (index: number) => void;
  /** The button displayed on the right */
  primaryAction?: React.ReactElement;
  /** The secondary button displayed on the left */
  secondaryAction?: React.ReactElement;
}

const Carousel: React.FunctionComponent<CarouselProps> = ({
  children,
  className,
  currentSlide = 0,
  dataTest,
  onPageChange,
  primaryAction,
  secondaryAction,
}: CarouselProps) => {
  const classes = cx('ids-carousel', className, {
    'ids-carousel--has-actions': secondaryAction || secondaryAction,
  });

  const childArray = React.Children.toArray(children);

  const handleOnPageChange = (index: number): void => {
    if (onPageChange) {
      onPageChange(index);
    }
  };

  return (
    <div className={classes} data-test={dataTest}>
      <div className="ids-carousel__slides">
        <div className="ids-carousel__slide">{childArray[currentSlide]}</div>
      </div>
      <div className="ids-carousel__nav-bar">
        <div className="ids-carousel__action-wrapper">
          {secondaryAction &&
            React.cloneElement(secondaryAction, {
              className: 'ids-carousel__secondary-action',
            })}
        </div>
        <div className="ids-carousel__bullet-list">
          {React.Children.map(childArray, (_child, index) => {
            const bulletClasses = cx('ids-carousel__bullet', {
              'ids-carousel__bullet--active': index === currentSlide,
            });
            return (
              <button
                className={bulletClasses}
                onClick={() => handleOnPageChange(index)}
                title={`Page ${index}`}
                aria-label={`Page ${index}`}
              />
            );
          })}
        </div>

        <div className="ids-carousel__action-wrapper">
          {primaryAction &&
            React.cloneElement(primaryAction, {
              className: 'ids-carousel__primary-action',
            })}
        </div>
      </div>
    </div>
  );
};

export default Carousel;
