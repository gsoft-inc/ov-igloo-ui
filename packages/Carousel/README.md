# Carousel

A carousel allows users to see and navigate the steps within a flow.

<Example is="custom" />

<ReferenceLinks is="custom" />

## Installation

To install `@igloo-ui/carousel` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/carousel
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/carousel
```

## Usage

Then to use the component in your code just import it!

```jsx
import Button from '@igloo-ui/button';
import Carousel from '@igloo-ui/carousel';

const SLIDE_NUM = 3;
const [selected, setSelected] = React.useState(0);
const [secondaryActionText, setSecondaryActionText] = React.useState('Cancel');
const [primaryActionText, setPrimaryActionText] = React.useState('Next');

const handlePageChange = (index) => {
  if (index < SLIDE_NUM - 1) {
    setPrimaryActionText('Next');
  } else {
    setPrimaryActionText('Done');
  }
  if (index > 0) {
    setSecondaryActionText('Prev');
  } else {
    setSecondaryActionText('Cancel');
  }
  setSelected(index);
};

const handlePrimaryActionClick = () => {
  if (selected < SLIDE_NUM - 1) {
    handlePageChange(selected + 1);
  }
};

const handleSecondaryActionClick = () => {
  if (selected > 0) {
    handlePageChange(selected - 1);
  }
};

<Carousel
  onPageChange={handlePageChange}
  currentSlide={selected}
  primaryAction={
    <Button appearance={'primary'} onClick={handlePrimaryActionClick}>
      {primaryActionText}
    </Button>
  }
  secondaryAction={
    <Button appearance={'ghost'} onClick={handleSecondaryActionClick}>
      {secondaryActionText}
    </Button>
  }
>
  <div style={{ background: 'pink', padding: '4rem' }}>Slide 1</div>
  <div style={{ background: 'lightBlue', padding: '4rem' }}>Slide 2</div>
  <div style={{ background: 'lightGreen', padding: '4rem' }}>Slide 3</div>
</Carousel>;
```
