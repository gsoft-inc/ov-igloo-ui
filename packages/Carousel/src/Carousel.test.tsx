/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen } from '@testing-library/react';

import Carousel from './Carousel';

const setup = (props = {}) => {
  return render(
    <Carousel dataTest="ids-carousel" {...props}>
      <div style={{ background: 'pink', padding: '4rem' }}>Slide 1</div>
      <div style={{ background: 'lightBlue', padding: '4rem' }}>Slide 2</div>
      <div style={{ background: 'lightGreen', padding: '4rem' }}>Slide 3</div>
    </Carousel>
  );
};

describe('Carousel', () => {
  test('It should render without errors', () => {
    setup();
    const wrapper = screen.getByTestId('ids-carousel');
    expect(wrapper).toBeInTheDocument();
  });

  test('It should render a snapshot', () => {
    const { asFragment } = setup();
    expect(asFragment()).toMatchSnapshot();
  });
});
