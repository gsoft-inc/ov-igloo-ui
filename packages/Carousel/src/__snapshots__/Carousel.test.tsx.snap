// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel It should render a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-carousel"
    data-test="ids-carousel"
  >
    <div
      class="ids-carousel__slides"
    >
      <div
        class="ids-carousel__slide"
      >
        <div
          style="background: pink; padding: 4rem;"
        >
          Slide 1
        </div>
      </div>
    </div>
    <div
      class="ids-carousel__nav-bar"
    >
      <div
        class="ids-carousel__action-wrapper"
      />
      <div
        class="ids-carousel__bullet-list"
      >
        <button
          aria-label="Page 0"
          class="ids-carousel__bullet ids-carousel__bullet--active"
          title="Page 0"
        />
        <button
          aria-label="Page 1"
          class="ids-carousel__bullet"
          title="Page 1"
        />
        <button
          aria-label="Page 2"
          class="ids-carousel__bullet"
          title="Page 2"
        />
      </div>
      <div
        class="ids-carousel__action-wrapper"
      />
    </div>
  </div>
</DocumentFragment>
`;
